### PR TITLE
8281946: VM.native_memory should report size of shareable memory

### PR DESCRIPTION
**Pull request by matias9927**
The native memory reporting has been improved to also include the amount of shareable (read only) data. Verified with test tiers 1,2, and 4.

Sample output:
```
Total: reserved=33890498670, committed=2360907886
       malloc: 75159662 #26118
       mmap:   reserved=33815339008, committed=2285748224
.............
-        Shared class space (reserved=16777216, committed=13545472, readonly=8315472)
                            (mmap: reserved=16777216, committed=13545472) 
 
-               Arena Chunk (reserved=441792, committed=441792)
                            (malloc=441792) 
```

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8281946](https://bugs.openjdk.org/browse/JDK-8281946): VM.native_memory should report size of shareable memory

### Reviewers
 * [Thomas Stuefe](https://openjdk.org/census#stuefe) (@tstuefe - **Reviewer**) ⚠️ Review applies to [17874d97](https://git.openjdk.org/jdk/pull/11401/files/17874d974801971f71ad0f68798fc05daeb66a69)
 * [Ioi Lam](https://openjdk.org/census#iklam) (@iklam - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/11401/head:pull/11401` \
`$ git checkout pull/11401`

Update a local copy of the PR: \
`$ git checkout pull/11401` \
`$ git pull https://git.openjdk.org/jdk pull/11401/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 11401`

View PR using the GUI difftool: \
`$ git pr show -t 11401`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/11401.diff">https://git.openjdk.org/jdk/pull/11401.diff</a>

</details>

### CODE REVIEW

**Review by tstuefe — COMMENTED** [review]
Hi Matias,

Nice, this is useful! Small nits, but nothing major.

About the terminology: Ioi did suggest "shareable", whereas you use "readonly". IMHO "shareable" is a bit clearer, and also more correct since on Windows, that memory is not readonly. Also, you can have read-only mappings that are not shareable, and rw mappings that are.

Long term, not for this RFE: it would be cool if we could track shareability inside NMT as property of a mapping info. Or even just query the OS if the mapping is shared or private. That way we can remove any flag-specific special handling from NMT, and become a lot more flexible (e.g. if someone shares memory for a different MEMFLAG). 

Cheers, Thomas

File: src/hotspot/share/services/memReporter.cpp [hunk 1]
```diff
@@ -105,6 +115,13 @@ void MemSummaryReporter::report() {
   size_t total_reserved_amount = total_malloced_bytes + total_mmap_reserved_bytes;
   size_t total_committed_amount = total_malloced_bytes + total_mmap_committed_bytes;
 
+  size_t read_only_bytes = 0;
+  FileMapRegion* r = FileMapInfo::current_info()->region_at(MetaspaceShared::ro);
+  // Region will be read-write on windows, otherwise this is a sanity check
+  if (!MetaspaceShared::use_windows_memory_mapping())
```
[the remark this answers is not in the document]
Comment (iklam) [reply src/hotspot/share/services/memReporter.cpp:121]: On Windows, the pages in the CDS archive are not shareable across processes (even the RO region is private to the current JVM process). Therefore, we shouldn't show them as shared. This condition should be changed to:

```
if (r->read_only()) {
   read_only_bytes = r->used();
}
```

Also, we can have up to two mapped CDS archives. For safety, NULL checks should be done on both:


```
read_only_bytes = get_read_only_size(FileMapInfo::current_info()); // static archive
read_only_bytes += get_read_only_size(FileMapInfo::dynamic_info()); // dynamic archive

size_t get_read_only_size(FileMapInfo* info) {
  if (info != NULL) {
    ....
   }
}
```

**Review by iklam — CHANGES_REQUESTED** [review]
Could you add a sample output to the PR summary? That way, people can see what it looks like without executing the new code in their heads :-)

File: test/hotspot/jtreg/runtime/NMT/SummarySanityCheck.java [hunk 2]
```diff
@@ -76,12 +79,33 @@ public static void main(String args[]) throws Exception {
...
+          long typeReserved = Long.parseLong(sharedMatcher.group("reserved"));
+          long typeShared = Long.parseLong(sharedMatcher.group("readonly"));
+
+          // Make sure reserved is always less or equals
+          if (typeCommitted > typeReserved) {
+            throwTestException("Committed (" + typeCommitted + ") was more than Reserved ("
+                + typeReserved + ") for mtType: " + typeMatcher.group("typename"));
+          }
+          // Make sure readonly is always less or equals
+          if (typeShared > typeReserved) {
```
[the remark this answers is not in the document]
Comment (matias9927) [reply test/hotspot/jtreg/runtime/NMT/SummarySanityCheck.java:100]: I believe you are correct!

**Comment by tstuefe** [comment]
> 
> The choice to use readonly over shareable was another suggestion by @iklam. Here is some rationale: "Other mmaped pages can be shareable if they are writable but has not been modified by the JVM process (ie copy-on-write pages)"

What. Why? @iklam ?

Writeable and shareable are orthogonal properties. You can have every permutation. And in the JDK we have:
- RO + shared: CDS, text segments, ...
- RW + shared: things like perfmem at least, but possibly other things too.
- RO + private: guard pages, ...
- RW + private: normal memory.

Here, we want to tell the user if something is shared: if the memory footprint is shouldered by several processes or by this process alone. The user does not really care about writeability, that's incidental. Therefore I think we should name it sharable.

**Comment by iklam** [comment]
> > The choice to use readonly over shareable was another suggestion by @iklam. Here is some rationale: "Other mmaped pages can be shareable if they are writable but has not been modified by the JVM process (ie copy-on-write pages)"
> 
> I still don't get it... @iklam ?
> 
> Writeable and shareable are orthogonal properties. You can have every permutation. And in the JDK we have:
> 
> * RO + shared: CDS ro segments, text segments, ...
> * RW + shared: things like perfmem at least, but possibly other things too.
> * RO + private: guard pages,  ...
> * RW + private: normal memory.
> 
> Here, we want to tell the user if something is shared: if the memory footprint is shouldered by several processes or by this process alone. The user does not really care about writeability, that's incidental. Therefore I think we should name it sharable.

I agree that the user might be more interested in shareability than writeability. However, that's not something that we can accurately report.

Regarding the ` Shared class space` part of the output, it covers only the `MetaspaceShared::ro` and `MetaspaceShared::rw` sections. Part of the `rw` sections may be shareable across processes because they are not written into during runtime. However, we cannot tell which pages are shared (unless we go through the `/proc/self/smaps` file, which would be too complicated for this RFE **)

If we report only the `ro` section as `shareable`, I think that's not accurate.

I see your point that adding `readonly` to these two lines is problematic, as we don't include the `PROT_READ` only mmap pages that are used for stackguards, for example.

```
Total: reserved=33890498670, committed=2360907886, readonly=8315472
       mmap:   reserved=33815339008, committed=2285748224, readonly=8315472
```

So for the output to be accurate, I think we should only change this one line, and leave the above two lines unchanged.

```
-        Shared class space (reserved=16777216, committed=13545472, readonly=8315472)
```

** Perhaps in a future RFE, we could go through the smaps file and report something like this for all the `mmap` lines:

```
       mmap:   reserved=33815339008, committed=2285748224, shared=12345
```

**Comment by tstuefe** [comment]
> > > The choice to use readonly over shareable was another suggestion by @iklam. Here is some rationale: "Other mmaped pages can be shareable if they are writable but has not been modified by the JVM process (ie copy-on-write pages)"
> > 
> > 
> > I still don't get it... @iklam ?
> > Writeable and shareable are orthogonal properties. You can have every permutation. And in the JDK we have:
> > 
> > * RO + shared: CDS ro segments, text segments, ...
> > * RW + shared: things like perfmem at least, but possibly other things too.
> > * RO + private: guard pages,  ...
> > * RW + private: normal memory.
> > 
> > Here, we want to tell the user if something is shared: if the memory footprint is shouldered by several processes or by this process alone. The user does not really care about writeability, that's incidental. Therefore I think we should name it sharable.
> 
> I agree that the user might be more interested in shareability than writeability. However, that's not something that we can accurately report.

Okay, I think I get it.

We map CDS with MAP_PRIVATE, either ro or rw. I thought we map with MAP_SHARED, and then just don't write to it.

We don't do anything explicit to share apart from mapping to the same address so we don't have to do relocations. We count on mapped private memory we don't write to being shared automatically by the kernel with other processes. But there is no easy way to observe the effect unless we look at /proc/pid/smaps Shared_Clean.

> 
> Regarding the ` Shared class space` part of the output, it covers only the `MetaspaceShared::ro` and `MetaspaceShared::rw` sections. Part of the `rw` sections may be shareable across processes because they are not written into during runtime. However, we cannot tell which pages are shared (unless we go through the `/proc/self/smaps` file, which would be too complicated for this RFE **)
> 
> If we report only the `ro` section as `shareable`, I think that's not accurate.
> 
> I see your point that adding `readonly` to these two lines is problematic, as we don't include the `PROT_READ` only mmap pages that are used for stackguards, for example.

Yes, its inaccurate either way. Because we can share a lot more than CDS RO sections. I think the biggest effect comes from sharing the text segments.

> 
> ```
> Total: reserved=33890498670, committed=2360907886, readonly=8315472
>        mmap:   reserved=33815339008, committed=2285748224, readonly=8315472
> ```
> 
> So for the output to be accurate, I think we should only change this one line, and leave the above two lines unchanged.
> 
> ```
> -        Shared class space (reserved=16777216, committed=13545472, readonly=8315472)
> ```
> 

I like this better too.

> ** Perhaps in a future RFE, we could go through the smaps file and report something like this for all the `mmap` lines:
> 
> ```
>        mmap:   reserved=33815339008, committed=2285748224, shared=12345
> ```

Sure, that would be possible. Note though that since we only report in NMT for mmap segments NMT knows about, the total number can always be misleading (as the others). E.g. we would not see sharing from text segments.

**Comment by iklam** [comment]
> We map CDS with MAP_PRIVATE, either ro or rw. I thought we map with MAP_SHARED, and then just don't write to it.

That's correct. The `MetaspaceShared::ro` region is usually mmapped with `flags=MAP_PRIVATE` with `prot=PROT_READ`

https://github.com/openjdk/jdk/blob/ea83cb960d07ffa9384aad6a1e2a0233e3ebbdd1/src/hotspot/os/linux/os_linux.cpp#L4930-L4940

**Comment by ashu-mehra** [comment]
@iklam not meant to stall this PR but I have wondered earlier why do we rely on OS's cow mechanism to make ro section shareable? As far as I know cow mapping for MAP_PRIVATE is not mandated by posix. We should instead just use MAP_SHARED flag for ro sections. That also makes it explicit that the section is meant to be shared.

**Comment by ashu-mehra** [comment]
> We should instead just use MAP_SHARED flag for ro sections.

To clarify, if the ro sections don't require relocation, use MAP_SHARED, else use MAP_PRIVATE.

**Comment by matias9927** [comment]
Since there is no guarantee that the readonly field will be accurate, the test was adjusted to allow "readonly=..." to be included in the output but only for "Shared class space" so long as the RO region is actually read only.

**Comment by iklam** [comment]
> > We should instead just use MAP_SHARED flag for ro sections.
> 
> To clarify, if the ro sections don't require relocation, use MAP_SHARED, else use MAP_PRIVATE.

I am not sure why we use MAP_PRIVATE + PROT_READ for CDS. Maybe it just happened to work so we end up doing it that way. I filed https://bugs.openjdk.org/browse/JDK-8298216 , quoting your suggestions from the previous comments.

File: src/hotspot/share/services/memReporter.cpp [hunk 3]
```diff
@@ -39,10 +39,28 @@ size_t MemReporterBase::committed_total(const MallocMemory* malloc, const Virtua
...
+  if (info!= NULL) {
+    FileMapRegion* r = info->region_at(MetaspaceShared::ro);
+    if (r->read_only()) {
+      total = r->used();
+    }
+  }
+  return total;
+}
+
+void MemReporterBase::print_total(size_t reserved, size_t committed, size_t read_only) const {
```
Comment (tstuefe) [inline src/hotspot/share/services/memReporter.cpp:55]: Since we now only print readonly for one tag in one report, we can keep this patch much smaller. The lLess complex the better. print_total gets called from several places that have nothing to do with CDS readonly.

My suggestion would be to CDS-specific stuff directly in `MemSummaryReporter::report_summary_of_type` since only there we only want the "Shared class" line in the summary report to show readonly. My proposal would be:

```
--- a/src/hotspot/share/services/memReporter.cpp
+++ b/src/hotspot/share/services/memReporter.cpp
@@ -214,7 +214,11 @@ void MemSummaryReporter::report_summary_of_type(MEMFLAGS flag,
     outputStream* out   = output();
     const char*   scale = current_scale();
     out->print("-%26s (", NMTUtil::flag_to_name(flag));
-    flag == mtClassShared ? print_total(reserved_amount, committed_amount, read_only_bytes) : print_total(reserved_amount, committed_amount);
+    print_total(reserved_amount, committed_amount);
+    if (flag == mtClassShared) {
+      output()->print(", readonly=" SIZE_FORMAT "%s",
+                      amount_in_current_scale(read_only_bytes), scale);
+    }
```

File: src/hotspot/share/services/memReporter.cpp [hunk 4]
```diff
@@ -161,12 +181,12 @@ void MemSummaryReporter::report() {
...
     VirtualMemory* virtual_memory = _vm_snapshot->by_type(flag);
 
-    report_summary_of_type(flag, malloc_memory, virtual_memory);
+    report_summary_of_type(flag, malloc_memory, virtual_memory, read_only_bytes);
   }
 }
 
 void MemSummaryReporter::report_summary_of_type(MEMFLAGS flag,
-  MallocMemory*  malloc_memory, VirtualMemory* virtual_memory) {
+  MallocMemory*  malloc_memory, VirtualMemory* virtual_memory, size_t read_only_bytes) {
```
Comment (tstuefe) [inline src/hotspot/share/services/memReporter.cpp:189]: You can actually move the retrieval of CDS read_only_bytes into this function now. No need to expand the interface. This function already retrieves a number of things, see e.g. thead stack printing, so there is a precedence.

File: test/hotspot/jtreg/runtime/NMT/SummarySanityCheck.java [hunk 5]
```diff
@@ -64,8 +64,8 @@ public static void main(String args[]) throws Exception {
     long totalCommitted = 0, totalReserved = 0;
     long totalCommittedSum = 0, totalReservedSum = 0;
 
-    // Match '- <mtType> (reserved=<reserved>KB, committed=<committed>KB)
-    Pattern mtTypePattern = Pattern.compile("-\\s+(?<typename>[\\w\\s]+)\\(reserved=(?<reserved>\\d+)KB,\\scommitted=(?<committed>\\d+)KB\\)");
+    // Match '- <mtType> (reserved=<reserved>KB, committed=<committed>KB) and some times readonly=<readonly>KB
+    Pattern mtTypePattern = Pattern.compile("-\\s+(?<typename>[\\w\\s]+)\\(reserved=(?<reserved>\\d+)KB,\\scommitted=(?<committed>\\d+)KB((,\\sreadonly=(?<readonly>\\d+)KB)|)\\)");
```
Comment (tstuefe) [inline test/hotspot/jtreg/runtime/NMT/SummarySanityCheck.java:68]: You can simplify this, since the only line where we want to see readonly is "Shared class space". That makes the test also more strict, since we do not want to see readonly on any other line.

**Review by tstuefe — CHANGES_REQUESTED** [review]
Hi Matias,

see remarks inline. Since the goal is only printing readonly on the "Shared class" line, the patch can be simplified. I am concerned with keeping the patch as simple as possible because reporting code is already quite complex and brittle. I'd like to make future revamps as easy as possible.

Cheers, Thomas

**Comment by tstuefe** [comment]
> > > We should instead just use MAP_SHARED flag for ro sections.
> > 
> > 
> > To clarify, if the ro sections don't require relocation, use MAP_SHARED, else use MAP_PRIVATE.
> 
> I am not sure why we use MAP_PRIVATE + PROT_READ for CDS. Maybe it just happened to work so we end up doing it that way. I filed https://bugs.openjdk.org/browse/JDK-8298216 , quoting your suggestions from the previous comments.

I commented in JBS. I'm not convinced this is a good idea.

File: src/hotspot/share/services/memReporter.hpp [hunk 6]
```diff
@@ -25,6 +25,7 @@
 #ifndef SHARE_SERVICES_MEMREPORTER_HPP
 #define SHARE_SERVICES_MEMREPORTER_HPP
 
+#include "cds/filemap.hpp"
```
Comment (tstuefe) [inline src/hotspot/share/services/memReporter.hpp:28]: Not necessary in the header. Please move to the cpp file.

File: src/hotspot/share/services/memReporter.hpp [hunk 7]
```diff
@@ -77,6 +78,7 @@ class MemReporterBase : public StackObj {
   // Calculate total reserved and committed amount
   size_t reserved_total(const MallocMemory* malloc, const VirtualMemory* vm) const;
   size_t committed_total(const MallocMemory* malloc, const VirtualMemory* vm) const;
+  size_t readonly_total(FileMapInfo* info) const;
```
Comment (tstuefe) [inline src/hotspot/share/services/memReporter.hpp:81]: Not necessary

File: src/hotspot/share/services/memReporter.cpp [hunk 8]
```diff
@@ -170,6 +171,12 @@ void MemSummaryReporter::report_summary_of_type(MEMFLAGS flag,
 
   size_t reserved_amount  = reserved_total (malloc_memory, virtual_memory);
   size_t committed_amount = committed_total(malloc_memory, virtual_memory);
+  // There can be up to two CDS archives which can contain readonly data. On Windows, pages are not
+  // shareable so the RO region may not actually be read only
+  size_t read_only_bytes;
+  FileMapInfo::current_info() != nullptr ? read_only_bytes = FileMapInfo::current_info()->readonly_total() : read_only_bytes = 0; // static archive
+  if (FileMapInfo::dynamic_info() != nullptr)
+    read_only_bytes += FileMapInfo::dynamic_info()->readonly_total(); // dynamic archive
```
Comment (tstuefe) [inline src/hotspot/share/services/memReporter.cpp:179]: Still does not belong here in this detail... NMT should not have to know that there are multiple FileMapInfo; it should not even have to know what a FileMapInfo is.

Proposal: Merge this into `FileMapInfo::readonly_total()`, make `FileMapInfo::readonly_total()` static, adapt comment (comment above about Windows details belongs atop of the prototype). And here, we just want to have a "const size_t readonly = FileMapInfo::readonly_total".

Ideally, it would be cool had CDS an outward facing minimal header (e.g. cdsStatistics.hpp) with APIs like these (e.g. `static size_t CDS::readonly_total()`. Because we still need to include the rather fat header fileMapInfo.hpp.

**Review by tstuefe — COMMENTED** [review]
Hi Matias,

we are getting there... see comment inline.

Cheers, Thomas

File: src/hotspot/share/services/memReporter.cpp (hunk 8 above)
Comment (matias9927) [reply src/hotspot/share/services/memReporter.cpp]: Thank you for the feedback! I agree that NMT should be blind to these features but I'm not entirely sure the best way to pass on the readonly amount without at least having FileMapInfo. Maybe there should be an RFE for the minimal header and then some cleanup?

File: src/hotspot/share/services/memReporter.cpp [hunk 9]
```diff
@@ -42,7 +43,8 @@ size_t MemReporterBase::committed_total(const MallocMemory* malloc, const Virtua
 void MemReporterBase::print_total(size_t reserved, size_t committed) const {
   const char* scale = current_scale();
   output()->print("reserved=" SIZE_FORMAT "%s, committed=" SIZE_FORMAT "%s",
-    amount_in_current_scale(reserved), scale, amount_in_current_scale(committed), scale);
+                  amount_in_current_scale(reserved), scale,
+                  amount_in_current_scale(committed), scale);
```
Comment (iklam) [inline src/hotspot/share/services/memReporter.cpp:47]: This change is unrelated and should be reverted.

File: test/hotspot/jtreg/runtime/NMT/SummarySanityCheck.java [hunk 10]
```diff
@@ -85,6 +85,15 @@ public static void main(String args[]) throws Exception {
           long typeCommitted = Long.parseLong(typeMatcher.group("committed"));
           long typeReserved = Long.parseLong(typeMatcher.group("reserved"));
 
+          if(typeMatcher.group("readonly") != null) {
+            long typeReadOnly = Long.parseLong(typeMatcher.group("readonly"));
```
Comment (iklam) [inline test/hotspot/jtreg/runtime/NMT/SummarySanityCheck.java:89]: Add a space behind the "if".

Also, as @tstuefe suggested, if "readonly" is present in the outline, check that the line also contains "Shared class space", since no other lines should contain "readonly".

**Review by iklam — CHANGES_REQUESTED** [review]

File: src/hotspot/share/services/memReporter.cpp [hunk 11]
```diff
@@ -195,6 +201,10 @@ void MemSummaryReporter::report_summary_of_type(MEMFLAGS flag,
     const char*   scale = current_scale();
     out->print("-%26s (", NMTUtil::flag_to_name(flag));
     print_total(reserved_amount, committed_amount);
+    if (flag == mtClassShared) {
+      output()->print(", readonly=" SIZE_FORMAT "%s",
+                      amount_in_current_scale(read_only_bytes), scale);
+    }
```
Comment (iklam) [inline src/hotspot/share/services/memReporter.cpp:207]: Two days ago, I made a request for change in here but that seems to have disappeared :-(

The read_only_bytes is relevant to mtClassShared only, so the code from lines 174-178 should be moved inside this "if" block.

**Review by iklam — APPROVED** [review]
LGTM!

**Comment by matias9927** [comment]
Thank you for your patience and reviews @iklam and @tstuefe! I re-ran tier1 tests and the macos-x64 tests pass, so I think we can safely ignore GHA here.
/integrate

**Comment by iklam** [comment]
@matias9927 Thank you for taking our suggestions and going through many revisions. I think the result is good.
/sponsor
